### PR TITLE
[SPARK-11926][SQL] unify GetStructField and GetInternalRowField

### PR DESCRIPTION
**Pull request by cloud-fan**

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala [hunk 1]
```diff
@@ -97,18 +97,15 @@ object ExtractValue {
  * Returns the value of fields in the Struct `child`.
  *
  * No need to do type checking since it is handled by [[ExtractValue]].
- * TODO: Unify with [[GetInternalRowField]], remove the need to specify a [[StructField]].
  */
-case class GetStructField(child: Expression, field: StructField, ordinal: Int)
+case class GetStructField(child: Expression, ordinal: Int, name: Option[String] = None)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Maybe comment that the `name` is only for debugging?

**Comment by marmbrus** [comment]
Thanks for cleaning this up!  LGTM pending tests.

**Comment by cloud-fan** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala (hunk 1 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: It's not for debugging, but to keep case preserving. For example, if we try to get field `yEAr` from `<year: int, month: int>`, we should pass `yEAr` to `GetStructField`. And it's also consistent with the original behavior.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypeExtractors.scala]: Oh, then document that :)

**Comment by marmbrus** [comment]
Merging to master and 1.6.

**Pull request closed without merging**